helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -0,0 +1,982 @@
+                    rebalanceType));
+          }
+          continue;
+        }
+      }
+
+      if (throttleController.shouldThrottleForResource(rebalanceType, resourceName)) {
+        throttledMessages.add(msg);
+        if (logger.isDebugEnabled()) {
+          LogUtil.logDebug(logger, _eventId,
+              String.format("Message: %s throttled in resource: %s with type: %s", msg, resourceName,
+                  rebalanceType));
+        }
+        continue;
+      }
+      String instance = msg.getTgtName();
+      if (throttleController.shouldThrottleForInstance(rebalanceType,instance)) {
+        throttledMessages.add(msg);
+        if (logger.isDebugEnabled()) {
+          LogUtil.logDebug(logger, _eventId,
+              String.format("Message: %s throttled in instance %s in resource: %s with type: %s", instance, msg, resourceName,
+                  rebalanceType));
+        }
+        continue;
+      }
+      throttleController.chargeInstance(rebalanceType, instance);
+      throttleController.chargeResource(rebalanceType, resourceName);
+      throttleController.chargeCluster(rebalanceType);
+      logger.trace("throttleControllerstate->{} after charge load msg: {}", throttleController, msg);
+    }
+  }
+
+   // ------------------ utilities ---------------------------
+  /**
+   * POJO that maps resource name to its priority represented by an integer.
+   */
+  private static class ResourcePriority {
+    private String _resourceName;
+    private int _priority;
+
+    ResourcePriority(String resourceName, Integer priority) {
+      _resourceName = resourceName;
+      _priority = priority;
+    }
+
+    public int compareTo(ResourcePriority resourcePriority) {
+      return Integer.compare(_priority, resourcePriority._priority);
+    }
+
+    public String getResourceName() {
+      return _resourceName;
+    }
+
+    public void setPriority(String priority) {
+      try {
+        _priority = Integer.parseInt(priority);
+      } catch (Exception e) {
+        logger.warn(
+            String.format("Invalid priority field %s for resource %s", priority, _resourceName));
+      }
+    }
+  }
+
+  private static class ResourcePriorityComparator implements Comparator<ResourcePriority> {
+    @Override
+    public int compare(ResourcePriority priority1, ResourcePriority priority2) {
+      return priority2.compareTo(priority1);
+    }
+  }
+
+  // compare message for throttling, note, all these message are of type state_transition: how about upward, downward?
+  // recovery are all upward
+  // 1) toState priority (toTop is higher than toSecond)
+  // 2) same toState, the message classification time, the less required toState meeting minActive requirement has higher priority
+  // 3) Higher priority for the partition of messages with fewer replicas with states matching with bestPossible ??? do we need this one
+  private static class MessageThrottleComparator implements Comparator<Message> {
```

Comment:
As you said, if there is no diff from this comparator and another one. Please merge them. They are very complicated. It will be hard for reader to understand the both. I will review the logic in detail once you merged.

Reply:
See my comment before. Let us chat about this one.